src/components/tl/mlx5/mcast/tl_mlx5_mcast_team.c

```diff
@@ -128,3 +298,241 @@ ucc_status_t ucc_tl_mlx5_mcast_coll_setup_comm_resources(ucc_tl_mlx5_mcast_coll_
+                }
+
+                ucc_assert(comm->event != NULL);
+
+                /* at this point, rank 0 has joined mcast group */
+                tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_GRP_JOIN_READY;
+            }
+
+            case TL_MLX5_TEAM_STATE_MCAST_GRP_JOIN_READY:
+            {
+                /* now rank 0 bcast the lid/gid to other processes */
+                mgid_s = sizeof(tmp.param.ud.ah_attr.grh.dgid);
+                mlid_s = sizeof(tmp.param.ud.ah_attr.dlid);
+                data   = ucc_malloc(sizeof(ucc_status_t) + mgid_s + mlid_s);
+                if (!data) {
+                    tl_error(comm->lib, "unable to allocate memory for group setup info");
+                    return UCC_ERR_NO_MEMORY;
+                }
+
+                comm->group_setup_info = data;
+                memset(data, 0, sizeof(ucc_status_t) + mgid_s + mlid_s);
+                memcpy((void*)((ptrdiff_t)data + sizeof(ucc_status_t)),
+                        &comm->event->param.ud.ah_attr.grh.dgid, mgid_s);
+                memcpy((void*)((ptrdiff_t)data + sizeof(ucc_status_t) + mgid_s),
+                        &comm->event->param.ud.ah_attr.dlid, mlid_s);
+                memcpy(data, &status, sizeof(ucc_status_t));
+
+                status = comm->bcast_post(comm->p2p_ctx, data, sizeof(ucc_status_t) + mgid_s +
+                                          mlid_s, 0, &comm->group_setup_info_req);
+                if (UCC_OK != status) {
+                    tl_error(comm->lib, "unable to post bcast for group setup info");
+                    ucc_free(comm->group_setup_info);
+                    if (comm->event) {
+                        rdma_ack_cm_event(comm->event);
+                    }
+                    return status;
+                }
+
+                comm->mcast_lid  = *((uint16_t*)((ptrdiff_t)data + sizeof(ucc_status_t) +
+                                     mgid_s));
+                comm->mcast_addr = net_addr;
+
+                memcpy((void*)&comm->mgid, (void*)((ptrdiff_t)data + sizeof(ucc_status_t)),
+                       mgid_s);
+
+                tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_GRP_BCAST_POST;
+            }
+
+            case TL_MLX5_TEAM_STATE_MCAST_GRP_BCAST_POST:
+            {
+                /* rank 0 polls bcast request and wait for its completion */
+                status = comm->bcast_test(comm->group_setup_info_req);
+                if (UCC_OK != status) {
+                    /* bcast is not completed yet */
+                    return status;
+                }
+
+                ucc_free(comm->group_setup_info);
+                if (comm->event) {
+                    rdma_ack_cm_event(comm->event);
+                }
+
+                /* setup of the rest of the mcast resources */
+                status = ucc_tl_mlx5_mcast_coll_setup_comm_resources(comm);
+                if (UCC_OK != status) {
+                    return status;
```

Comment:
free `data`?

Reply:
it is being freed by ````ucc_free(comm->group_setup_info);````

Reply:
data pointer is same as comm->group_setup_info, so it should have been already freed before calling ucc_tl_mlx5_mcast_coll_setup_comm_resources()